### DISCUSSION TITLE
`text` in `insert().values(data)` vs execute(`insert()`, data)

### DISCUSSION THREAD

**Discussion by CaselIT** [sqlalchemy/sqlalchemy#8532] [Usage Questions]
Opening a discussion since I'm not sure if this is expected or a bug

Currently it's possible to execute this
```py
import sqlalchemy as sa

e = sa.create_engine("postgresql://scott:tiger@localhost/test", echo=True)
metadata = sa.MetaData()
Foo = sa.Table("foo", metadata, sa.Column("id", sa.Integer))

with e.connect() as conn:
    metadata.create_all(conn)
    i_query = Foo.insert().values([{"id": sa.text("1")}, {"id": sa.text("2")}])
    conn.execute(i_query)
```

but trying to execute this will instead fail
```py
import sqlalchemy as sa

e = sa.create_engine("postgresql://scott:tiger@localhost/test", echo=True)
metadata = sa.MetaData()
Foo = sa.Table("foo", metadata, sa.Column("id", sa.Integer))

with e.connect() as conn:
    metadata.create_all(conn)
    conn.execute(
        Foo.insert(),
        [{"id": sa.text("1")}, {"id": sa.text("2")}],
    )
```
with the exception
```
sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) can't adapt type 'TextClause'
[SQL: INSERT INTO foo (id) VALUES (%(id)s)]
[parameters: [{'id': <sqlalchemy.sql.elements.TextClause object at 0x0000024FAC031AB0>}, {'id': <sqlalchemy.sql.elements.TextClause object at 0x0000024FAC031A50>}]]
(Background on this error at: https://sqlalche.me/e/20/f405)
```
since it seem that the test clause was not compiled

**Comment by zzzeek** [comment]
that's the expected behavior because parameters sent as the second argument to any .execute() method are parameters for a statement.  SQL expressions are not compiled on that side.

in most cases, this is generally intuitive.   with the INSERT, it looks a little weird once one becomes aware of values().     So towards this difference, we have a little bit of a note, though not that specifically so that could be improved, at https://docs.sqlalchemy.org/en/14/core/dml.html#sqlalchemy.sql.expression.Insert.values  , and in the new docs for ORM  bulk, I have very explicit compare-and-constrast language which you an see at https://docs.sqlalchemy.org/en/gerrit/4042/orm/queryguide/dml.html#orm-bulk-insert-with-per-row-sql-expressions and it's also in the table at the top.

**Reply by CaselIT** [reply]
The issue came up since it seems that at the moment we can't insert `array['[{"x": 1}]'::json]` as as python objects, since passing `[[{'x':1}]]` to a `Column(ARRAY(JSON))` will insert a two dimensional array with the json `{"x": 1}` so `text` is required.

I agree that the documentation could be improved

**Reply by zzzeek** [reply]
we support CASTS for this, shouldn't we have the CAST generated automatically?  does it work with asyncpg, where we should be doing that alrready?

**Reply by CaselIT** [reply]
I have not tried cast in values. With execute cast does not work, since it does as text

**Reply by zzzeek** [reply]
I meant, you would just send [{"x":1}] as the value directly and a ::array[json] would be produced automatically

**Reply by CaselIT** [reply]
That works, the issue is that the data is ambiguous when inserting a json array, since at the moment it will insert a two dimension sql array containing a json object.

Again I'm not sure it can do better, since the value is ambiguous
